File: conformance/drivers/driver.py
"""Conformance driver (Python). See conformance/run.py.

Usage: python3 driver.py <task.db> <scenario.json>
"""

import json
import sys

import mundane


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: driver.py <task.db> <scenario.json>", file=sys.stderr)
        return 2
    db_path, scn_path = sys.argv[1], sys.argv[2]
    with open(scn_path) as f:
        scenario = json.load(f)

    def wf(ctx):
        for op in scenario["operations"]:
            if op["op"] == "step":
                value = op.get("value")
                ctx.step(op["name"], lambda value=value: value)
            elif op["op"] == "sleep":
                ctx.sleep(op["name"], op["duration"])
            else:
                raise ValueError(f"unknown op: {op['op']}")

    mundane.run(db_path, wf)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: conformance/drivers/driver.js
// Conformance driver (TypeScript/Node). See conformance/run.py.
//
// Usage: node driver.js <task.db> <scenario.json>
// Requires the compiled @mundane/core (run `tsc -p .` in typescript/ first).

const fs = require("node:fs");
const { run } = require("../../typescript/dist/src/index.js");

async function main() {
  const [dbPath, scnPath] = process.argv.slice(2);
  if (!dbPath || !scnPath) {
    console.error("usage: node driver.js <task.db> <scenario.json>");
    process.exit(2);
  }
  const scenario = JSON.parse(fs.readFileSync(scnPath, "utf8"));
  await run(dbPath, async (ctx) => {
    for (const op of scenario.operations) {
      if (op.op === "step") {
        await ctx.step(op.name, async () => op.value ?? null);
      } else if (op.op === "sleep") {
        await ctx.sleep(op.name, op.duration);
      } else {
        throw new Error(`unknown op: ${op.op}`);
      }
    }
  });
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
